File: 074_MergeIntervals.py
"""
Topic: Intervals / Sorting
Problem: Merge Intervals
Description: Merge all overlapping intervals.
Time: O(n log n)
Space: O(n)
"""

from typing import List


def merge(intervals: List[List[int]]) -> List[List[int]]:
    if not intervals:
        return []
    intervals.sort(key=lambda x: x[0])
    res = [intervals[0]]
    for s, e in intervals[1:]:
        last_s, last_e = res[-1]
        if s <= last_e:
            res[-1][1] = max(last_e, e)
        else:
            res.append([s, e])
    return res


if __name__ == '__main__':
    print(merge([[1,3],[2,6],[8,10],[15,18]]))  # [[1,6],[8,10],[15,18]]
    print(merge([[1,4],[4,5]]))  # [[1,5]]

File: 074_MergeIntervals.cpp
// Topic: Intervals / Sorting
// Merge Intervals: given intervals, merge all overlapping intervals
// Time Complexity: O(n log n)
// Space Complexity: O(n) for output

#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> mergeIntervals(vector<vector<int>>& intervals)
{
    if (intervals.empty())
        return {};
    sort(intervals.begin(), intervals.end());
    vector<vector<int>> res;
    res.push_back(intervals[0]);
    for (int i = 1; i < intervals.size(); ++i)
    {
        if (intervals[i][0] <= res.back()[1])
            res.back()[1] = max(res.back()[1], intervals[i][1]);
        else
            res.push_back(intervals[i]);
    }
    return res;
}

int main()
{
    vector<vector<int>> a1 = {{1, 3}, {2, 6}, {8, 10}, {15, 18}};
    auto r1 = mergeIntervals(a1);
    for (auto& iv : r1) cout << "{" << iv[0] << "," << iv[1] << "} ";
    cout << "\n";  // {1,6} {8,10} {15,18}
    vector<vector<int>> a2 = {{1, 4}, {4, 5}};
    auto r2 = mergeIntervals(a2);
    for (auto& iv : r2) cout << "{" << iv[0] << "," << iv[1] << "} ";
    cout << "\n";  // {1,5}
    return 0;
}
